### DISCUSSION TITLE
RTSP SDP Parameters / H265 HDMI Extender Decoder

### DISCUSSION THREAD

**Discussion by donaldlaforce** [bluenviron/mediamtx#3505] [Questions]
### Question

Version: MediaMTX v1.8.3 on Debian 12

I came up with an interesting use of MediaMTX but I've hit a bit of a snag. 

Background:
I have been testing a h264/h265 media decoder: HDC-E5200K, rebranded as a "Monoprice Blackbird H.265 HDMI over IP Decoder/Receiver" https://www.monoprice.com/product?p_id=43625

The thought was that MediaMTX could be used as a software-based routing solution for many of these decoders.

These decoders immediately power on and search for an RTSP server to stream from at 192.168.10.10

The main issue that I am running into is that the SDP control parameters returned by MediaMTX causes the decoder to have a malformed RTSP request.

MediaMTX log:

```
2024/06/22 16:09:52 INF [RTSP] [conn 192.168.10.20:36228] opened
2024/06/22 16:09:52 INF [RTSP] [session 42aa0b00] created by 192.168.10.20:36228
2024/06/22 16:09:52 INF [RTSP] [conn 192.168.10.20:36228] closed: invalid path name: can contain only alphanumeric characters, underscore, dot, tilde, minus or slash (live/main/av_stream/rtsp://192.168.10.10:554/live/main/av_stream)
2024/06/22 16:09:52 INF [RTSP] [session 42aa0b00] destroyed: not in use
```

This is the SDP response from the official encoder:

```
root@debian:~# printf "DESCRIBE rtsp://192.168.10.10:554/live/main/av_stream RTSP/1.0\r\nCSeq: 1\r\nUser-Agent: RTSP Lib\r\nAccept: application/sdp\r\n\r\n" | nc 192.168.10.10 554
RTSP/1.0 200 OK
CSeq: 1
Date: Thu, 01 Jan 1970 00:00:56 GMT
Content-Type: application/sdp
Content-Length: 502

v=0
o=- 9487534896915416231 9487534896915442000 IN IP4 0.0.0.0
s=rtsp://192.168.10.10:554/live/main/av_stream
c=IN IP4 0.0.0.0
t=0 0
a=rtcp-unicast: reflection
m=video 0 RTP/AVP 98
c=IN IP4 0.0.0.0
b=AS:10240
a=recvonly
a=control:track1
a=framerate:30
a=rtpmap:98 H265/90000
a=fmtp:98 profile-level-id=4D0032; packetization-mode=1; sprop-parameter-sets=Z00AMpY1QEABg03AQEBAgA==,aO48gA==
m=audio 0 RTP/AVP 108
b=AS:128
c=IN IP4 0.0.0.0
a=recvonly
a=control:track2
a=rtpmap:108 WAV/48000/2
```

This is the SDP response from MediaMTX:

```
root@debian:~# printf "DESCRIBE rtsp://192.168.10.10:554/live/main/av_stream RTSP/1.0\r\nCSeq: 1\r\nUser-Agent: RTSP Lib\r\nAccept: application/sdp\r\n\r\n" | nc 192.168.10.10 554
RTSP/1.0 200 OK
CSeq: 1
Content-Base: rtsp://192.168.10.10:554/live/main/av_stream/
Content-Length: 332
Content-Type: application/sdp
Server: gortsplib

v=0
o=- 0 0 IN IP4 127.0.0.1
s=Session streamed with GStreamer
c=IN IP4 0.0.0.0
t=0 0
m=video 0 RTP/AVP 96
a=control:rtsp://192.168.10.10:554/live/main/av_stream/trackID=0
a=rtpmap:96 H264/90000
a=fmtp:96 packetization-mode=1; profile-level-id=42C028; sprop-parameter-sets=Z0LAKNoB4AiflwFqAgICgAAAAwCAAAAZR4wZUA==,aM48gA==
```

It looks like this is because the SDP control parameter consists of the entire media RTSP path rather than just the track name. I have gone through the documentation and config files, but haven't found any obvious ways to manipulate the SDP responses for troublesome clients.

**Comment by aler9** [comment]
Hello,

> It looks like this is because the SDP control parameter consists of the entire media RTSP path rather than just the track name

Your analysis is correct, in MediaMTX the control attribute contains an absolute path, while in other RTSP implementations you can find a relative one. We did this change in https://github.com/bluenviron/gortsplib/pull/210 because GStreamer was unable to merge together all the elements that make a RTSP path, that are the base URL inside `Content-Base` and the control attribute (`a=control:`). In particular, GStreamer is unable to append the control attribute after any query parameter. Therefore, we chose to merge all these elements on the server-side, resulting in an absolute URL.

If you need a different behavior, you can download gortsplib and remove these lines:

https://github.com/bluenviron/gortsplib/blob/cb3c527bc1a52dad0cba02ec56cc9d8d7e7e5f2a/server_conn.go#L46C1-L50C26

Then download MediaMTX, edit go.mod and add a link to your custom gortsplib:

```
replace github.com/bluenviron/gortsplib/v4 => ./path/to/gortsplib
```

Then compile the server:

```
go generate ./...
go run .
```

[a comment by an automated account is not shown]
